gRPC integration

c24t opened the pull request:
This is a port of [`grpcio-opentracing`](https://github.com/opentracing-contrib/python-grpc/), and borrows from [`opencensus-ext-grpc`](https://github.com/census-instrumentation/opencensus-python/tree/master/contrib/opencensus-ext-grpc). It adds client and server interceptors that wrap each request in a span and use the new context API to inject/extract the traceparent header to/from gRPC metadata.

 The PR includes two examples copied from [grpc-python](https://github.com/grpc/grpc/tree/master/examples/python/) itself:

- `helloworld` ([source](https://github.com/grpc/grpc/tree/master/examples/python/helloworld), [docs](https://grpc.io/docs/quickstart/python/)) which shows how to instrument the client and server in a basic unary-unary call pattern
- `route_guide` ([source](https://github.com/grpc/grpc/tree/master/examples/python/route_guide), [docs](https://grpc.io/docs/tutorials/basic/python/)) which includes examples for unary-unary, client-streaming, server-streaming, and bidirectional streaming

**Running the examples**

```
% cd ext/opentelemetry-ext-grpc/examples
% ./hello_world_server.py
Span(name="/helloworld.Greeter/SayHello", context=SpanContext(trace_id=0xdc68d08b983225bf3b86eabfc9a9f02c, span_id=0x12aa95fc91338970, trace_state={}), kind=SpanKind.SERVER, parent=<opentelemetry.trace.DefaultSpan object at 0x10943d190>, start_time=2020-03-25T00:23:31.605991Z, end_time=2020-03-25T00:23:31.606155Z)
```

```
% ./hello_world_client.py
Span(name="/helloworld.Greeter/SayHello", context=SpanContext(trace_id=0xdc68d08b983225bf3b86eabfc9a9f02c, span_id=0xb4766ea78457996c, trace_state={}), kind=SpanKind.CLIENT, parent=None, start_time=2020-03-25T00:23:31.574011Z, end_time=2020-03-25T00:23:31.606996Z)
Greeter client received: Hello, YOU!
```

Note that the client's trace ID gets propagated to the server.

You should see something similar running `./route_guide_server.py` and `route_guide_client.py`.

**Comparing to grpcio-opentracing**

88e94672 copied some files from [grpcio-opentracing](https://github.com/opentracing-contrib/python-grpc/). To see what changed between OpenTracing and OpenTelemetry, check the [diff against 88e94672](https://github.com/open-telemetry/opentelemetry-python/pull/476/files/88e946724c20fbcf36d74757aa0a1eecddcb986b..HEAD). The important files are `_client.py` and `_server.py`.

The `grpcext` package is copied from `grpcio-opentracing`. It implemented interceptors before `grpc-python` had proper support for them. `grpc-python` has since added interceptors, but `grpcio-opentracing` still uses its own implementation.

This PR still uses `grpcext`, but we should replace this with the official interceptor API in a follow up PR.

See for background:
- https://github.com/grpc-ecosystem/grpc-opentracing/pull/32
- https://github.com/grpc/grpc/issues/8767

**How to review**

Please ignore the generated `_pb2.py` and `_pb2_grpc.py` files. The meat of the PR is in `_client.py` and `_server.py`.

Try running the tests and examples.

**Still TODO**

- This PR only includes server-side unary-unary tests, add tests for both sides for all calling patterns.
- Add span attributes following the [semantic conventions in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/bc23ca12199836b565c867a393bf752d0fb25f04/specification/data-rpc.md).
- Use the new [gRPC interceptor API](https://github.com/grpc/grpc/blob/master/examples/python/interceptors/headers/greeter_server.py) instead of `grpcext`.

ocelotl commented on ext/opentelemetry-ext-grpc/examples/helloworld_pb2_grpc.py:
```diff
@@ -0,0 +1,46 @@
+import helloworld_pb2 as helloworld__pb2
+
+
+class GreeterStub(object):
```
It is not necessary to inherit from `object` in Python 3.

c24t replied:
Sorry @ocelotl, these are generated files.

ocelotl commented on ext/opentelemetry-ext-grpc/examples/helloworld_pb2_grpc.py:
```diff
@@ -0,0 +1,46 @@
+    raise NotImplementedError('Method not implemented!')
+
+
+def add_GreeterServicer_to_server(servicer, server):
```
Is the caps convention in this method name necessary?

ocelotl commented on ext/opentelemetry-ext-grpc/examples/helloworld_pb2_grpc.py:
```diff
@@ -0,0 +1,46 @@
+    """Sends a greeting
+    """
+    context.set_code(grpc.StatusCode.UNIMPLEMENTED)
+    context.set_details('Method not implemented!')
```
Double quotes for consistency

ocelotl commented on ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/client_interceptor.py:
```diff
@@ -0,0 +1,217 @@
+
+    def __next__(self):
+        try:
+            message = next(self._iterator)
```
Unused variable here

ocelotl commented on ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/client_interceptor.py:
```diff
@@ -0,0 +1,217 @@
+    def is_active(self):
+        return self._iterator.is_active()
+
+    def cancelled(self):
```
Can all these `NotImplementedError` be handled by `__getattr__` to avoid repeated code?

codeboten commented on .flake8:
```diff
@@ -17,3 +17,5 @@ exclude =
   __pycache__
   ext/opentelemetry-ext-jaeger/src/opentelemetry/ext/jaeger/gen/
   ext/opentelemetry-ext-jaeger/build/*
+  ext/opentelemetry-ext-grpc/examples/
```
This directory no longer exists.

codeboten commented:
Worth noting somewhere that there's a dependency to install `protobuf` in order to run the examples

codeboten commented on docs/examples/opentelemetry-example-app/src/opentelemetry_example_app/grpc/route_guide_client.py:
```diff
@@ -0,0 +1,160 @@
+
+from opentelemetry import trace
+from opentelemetry.ext.grpc import client_interceptor
+from opentelemetry.ext.grpc.grpcext import intercept_channel
```
Nit: `ext.grpc.grpcext` seems repetitive here

c24t replied:
I didn't move this because it's part of the gRPC "extensions" from grpcio-opentracing. That is, the code in `ext.grpc.grpcext` could be gRPC itself, code in `ext.grpc` is specific to OT.

When we switch to gRPC's built-in interceptors from https://github.com/grpc/grpc/issues/8767 we can get rid of this package.

codeboten commented on docs/examples/opentelemetry-example-app/src/opentelemetry_example_app/grpc/route_guide_server.py:
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python
+# Copyright 2020, OpenTelemetry Authors
```
Copyright should be updated to match https://github.com/open-telemetry/opentelemetry-python/pull/523

codeboten commented on ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/version.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__version__ = "0.5.dev0"
```
```suggestion
__version__ = "0.6.dev0"
```

c24t commented:
> Worth noting somewhere that there's a dependency to install `protobuf` in order to run the examples

Good call, I added it to the example app requirements and added a note in the example docstrings in d4b9ca0.

codeboten commented:
Did an initial pass on this, tried out the examples and it's looking pretty good! Got sidetracked troubleshooting the runtime context in the test, will finish the review today.

codeboten commented on ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/_server.py:
```diff
@@ -0,0 +1,224 @@
+        return self._servicer_context.set_details(details)
+
+
+# def _add_peer_tags(peer_str, tags):
```
Is this needed?

codeboten commented on ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/_server.py:
```diff
@@ -0,0 +1,224 @@
+# latter and updates the span accordingly.
+def _check_error_code(span, servicer_context, rpc_info):
+    if servicer_context.code != grpc.StatusCode.OK:
+        # span.set_tag("error", True)
```
Is this code still needed here?

johananl commented:
I've experimented with the code by running a Python gRPC server with a Go gRPC client (code is [here](https://github.com/johananl/otel-multi-language-demo/tree/e821cc2f47e9bcb792806c1aa2d0a3a1d0b12438/python/field) and [here](https://github.com/johananl/otel-multi-language-demo/tree/e821cc2f47e9bcb792806c1aa2d0a3a1d0b12438/go/frontend)). Seems to work nicely :+1:  Hope this helps.

codeboten commented:
@c24t I created a PR to your fork to remove the need to use ThreadLocalRuntimeContext in the server test. https://github.com/c24t/opentelemetry-python/pull/3

c24t commented:
Thanks @johananl and @codeboten!

toumorokoshi commented:
[grpc.txt](https://github.com/open-telemetry/opentelemetry-python/files/4405972/grpc.txt)

trying to figure out how to send PRs over, but here's a patch for tox directives.
